main.go

```diff
@@ -28,22 +33,48 @@ func main() {
 		URI:        cfg.MongoConfig.BindAddr,
 	}
 
-	if err := mongo.Init(); err != nil {
+	session, err := mongo.Init()
+	if err != nil {
 		log.ErrorC("Failed to initialise mongo", err, nil)
 		os.Exit(1)
 	}
 
-	router := mux.NewRouter()
-
-	s := server.New(cfg.BindAddr, router)
+	mongo.Session = session
 
 	log.Debug("listening...", log.Data{
 		"bind_address": cfg.BindAddr,
 	})
 
-	_ = api.CreateDatasetAPI(cfg.DatasetAPIURL, cfg.SecretKey, router, store.DataStore{Backend: mongo})
+	apiErrors := make(chan error, 1)
 
-	if err = s.ListenAndServe(); err != nil {
-		log.Error(err, nil)
+	api.CreateDatasetAPI(cfg.DatasetAPIURL, cfg.BindAddr, cfg.SecretKey, store.DataStore{Backend: mongo}, apiErrors)
+
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
+		log.Info(fmt.Sprintf("Shutdown with timeout: %s", cfg.ShutdownTimeout), nil)
+		ctx, cancel := context.WithTimeout(context.Background(), cfg.ShutdownTimeout)
+
+		api.Close(ctx)
```

Comment:
If the http server closes and then sends a signal to apiErrors channel, this will trigger a second close to the http server.

Reply:
This is true but not a problem as it will return in no time at all to then safely close other connections before service shuts down.

In the other scenario where there is an os.signal to terminate or interrupt then this call to close the api needs to take place.